lib/constructs/operation.js

```diff
@@ -79,14 +80,46 @@ class Operation {
     requires.merge(parameterConversions.requires);
     str += parameterConversions.body;
 
-    if (overloads.every(overload => conversions[overload.operation.idlType.idlType])) {
+    if (
+      this.ctx.options.skipWrapReturn ||
+      overloads.every(overload => conversions[overload.operation.idlType.idlType])
+    ) {
       str += `
         return ${callOn}.${implFunc}(${argsSpread});
       `;
     } else {
-      str += `
-        return utils.tryWrapperForImpl(${callOn}.${implFunc}(${argsSpread}));
-      `;
+      switch (this.generic) {
+        case "Promise": {
+          str += `
+            const res = ${callOn}.${implFunc}(${argsSpread});
+            return res && res.then ? res.then(utils.tryWrapperForImpl) : res;
```

Comment:
Let's make the code stricter. Only check `res` is null if the return type is nullable, and if it's not null (or the return type is not nullable) just do `res.then(utils.tryWrapperForImpl)`.

Ditto for the two blocks below.